File: pypi/src/spreen_pr/application.py
import re


class Application:
    """Derives a pull request title and labels from a topic branch name
    shaped ``{user}/{issue}/{category}/{summary}``, with the hotfix special
    case ``{user}/{issue}/hotfix/{category}/{summary}``."""

    EXPECTED_SHAPE = '{user}/{issue}/{category}/{summary}'

    @classmethod
    def run(cls, branch_name: str) -> str:
        return cls(branch_name=branch_name).title

    def __init__(self, branch_name: str) -> None:
        self._branch_name = branch_name
        self._segments = branch_name.split('/')
        self._validate()

    @property
    def title(self) -> str:
        labels = ''.join(f'[{label}]' for label in self.labels)
        return f'{labels} {self.topic}'

    @property
    def labels(self) -> list[str]:
        if self._is_hotfix():
            return [self._segments[-3].capitalize(), self._segments[-2]]
        return [self._segments[-2]]

    @property
    def topic(self) -> str:
        words = re.split(r'[-_]', self._segments[-1])
        return ' '.join(word.capitalize() for word in words)

    def to_dict(self) -> dict[str, object]:
        return {'title': self.title, 'labels': self.labels}

    # private

    def _is_hotfix(self) -> bool:
        return self._segments[-3] == 'hotfix'

    def _validate(self) -> None:
        if len(self._segments) >= 3 and all(self._segments):
            return
        raise ValueError(
            f'Branch name `{self._branch_name}` does not match the expected '
            f'`{self.EXPECTED_SHAPE}` shape')

File: rubygem/lib/spreen_pr/application.rb
# frozen_string_literal: true
# rbs_inline: enabled

module SpreenPr
  # Derives a pull request title and labels from a topic branch name shaped
  # `{user}/{issue}/{category}/{summary}`, with the hotfix special case
  # `{user}/{issue}/hotfix/{category}/{summary}`.
  class Application
    EXPECTED_SHAPE = '{user}/{issue}/{category}/{summary}' #: String

    # @rbs branch_name: String
    # @rbs return: String
    def self.run(branch_name:)
      new(branch_name:).title
    end

    # @rbs branch_name: String
    # @rbs return: void
    def initialize(branch_name:)
      @branch_name = branch_name
      @segments    = branch_name.split('/')
      validate!
    end

    # @rbs return: String
    def title
      "#{labels.map { |label| "[#{label}]" }.join} #{topic}"
    end

    # @rbs return: Array[String]
    def labels
      hotfix? ? [segments[-3].capitalize, segments[-2]] : [segments[-2]]
    end

    # @rbs return: String
    def topic
      segments.last.split(/[-_]/).map(&:capitalize).join("\s")
    end

    # @rbs return: Hash[Symbol, String | Array[String]]
    def to_h
      { title:, labels: }
    end

    private

    attr_reader :branch_name #: String
    attr_reader :segments #: Array[String]

    # @rbs return: bool
    def hotfix?
      segments[-3] == 'hotfix'
    end

    # @rbs return: void
    def validate!
      return if segments.size >= 3 && segments.none?(&:empty?)

      raise ArgumentError,
            "Branch name `#{branch_name}` does not match the expected `#{EXPECTED_SHAPE}` shape"
    end
  end
end
